File: IsraelHiking.Web/ios/App/App/Car/SpatialHelper.swift
import CoreLocation
import Foundation

/// Mirrors `SpatialHelper.kt`, and the web client's SpatialHelper: the spatial calculations the car
/// experience measures routes with, kept off `CLLocation.distance(from:)`, which allocates two
/// objects per call - a route with tens of thousands of points cannot afford that when it is
/// re-measured on every GPS fix.
enum SpatialHelper {

    /// Good enough for the local, flat plane calculations, the earth is not a perfect sphere anyway.
    static let metersPerLatitudeDegree = 111_320.0

    private static let earthRadiusMeters = 6_371_008.8

    /// How many meters a degree of longitude is worth at the given latitude.
    static func metersPerLongitudeDegree(atLatitude latitude: Double) -> Double {
        metersPerLatitudeDegree * cos(latitude * .pi / 180)
    }

    /// The distance in meters between two positions, measured on the sphere. Used for the distances
    /// along a route, which are measured once per route and have to hold over its whole length.
    static func distanceMeters(_ from: CLLocationCoordinate2D, _ to: CLLocationCoordinate2D) -> Double {
        let fromLatitude = from.latitude * .pi / 180
        let toLatitude = to.latitude * .pi / 180
        let latitudeDelta = (to.latitude - from.latitude) * .pi / 180
        let longitudeDelta = (to.longitude - from.longitude) * .pi / 180
        let a = sin(latitudeDelta / 2) * sin(latitudeDelta / 2)
            + cos(fromLatitude) * cos(toLatitude) * sin(longitudeDelta / 2) * sin(longitudeDelta / 2)
        return 2 * earthRadiusMeters * asin(min(a.squareRoot(), 1))
    }

    /// The bearing in degrees within [0, 360) from one position to another.
    static func bearingDegrees(_ from: CLLocationCoordinate2D, _ to: CLLocationCoordinate2D) -> Double {
        let fromLatitude = from.latitude * .pi / 180
        let toLatitude = to.latitude * .pi / 180
        let longitudeDelta = (to.longitude - from.longitude) * .pi / 180
        let y = sin(longitudeDelta) * cos(toLatitude)
        let x = cos(fromLatitude) * sin(toLatitude) - sin(fromLatitude) * cos(toLatitude) * cos(longitudeDelta)
        return (atan2(y, x) * 180 / .pi + 360).truncatingRemainder(dividingBy: 360)
    }

    /// Smallest absolute difference between two bearings, in degrees within [0, 180].
    static func angleDifference(_ a: Double, _ b: Double) -> Double {
        let difference = abs(a - b).truncatingRemainder(dividingBy: 360)
        return difference > 180 ? 360 - difference : difference
    }

    /// The distance in meters between a point and the closest place on a segment, measured on a
    /// plane laid around the point.
    static func perpendicularDistanceMeters(_ point: CLLocationCoordinate2D,
                                            _ from: CLLocationCoordinate2D,
                                            _ to: CLLocationCoordinate2D) -> Double {
        let metersPerLongitude = metersPerLongitudeDegree(atLatitude: point.latitude)
        let startX = (from.longitude - point.longitude) * metersPerLongitude
        let startY = (from.latitude - point.latitude) * metersPerLatitudeDegree
        let endX = (to.longitude - point.longitude) * metersPerLongitude
        let endY = (to.latitude - point.latitude) * metersPerLatitudeDegree
        let deltaX = endX - startX
        let deltaY = endY - startY
        let lengthSquared = deltaX * deltaX + deltaY * deltaY
        var projectionFactor = 0.0
        if lengthSquared > 0 {
            projectionFactor = min(max(-(startX * deltaX + startY * deltaY) / lengthSquared, 0), 1)
        }
        let x = startX + projectionFactor * deltaX
        let y = startY + projectionFactor * deltaY
        return (x * x + y * y).squareRoot()
    }

    /// Drops the points that fall within `toleranceMeters` of the line their neighbours draw
    /// (Ramer-Douglas-Peucker), keeping the corners. Iterative rather than recursive, since a
    /// recorded route can hold tens of thousands of points.
    static func simplify(_ points: [CLLocationCoordinate2D], toleranceMeters: Double) -> [CLLocationCoordinate2D] {
        guard points.count > 2 else { return points }
        var kept = [Bool](repeating: false, count: points.count)
        kept[0] = true
        kept[points.count - 1] = true
        var ranges = [(first: 0, last: points.count - 1)]
        while let range = ranges.popLast() {
            if range.last <= range.first + 1 {
                continue
            }
            var farthest = -1
            var farthestDistance = toleranceMeters
            for index in (range.first + 1)..<range.last {
                let distance = perpendicularDistanceMeters(
                    points[index], points[range.first], points[range.last])
                if distance > farthestDistance {
                    farthestDistance = distance
                    farthest = index
                }
            }
            if farthest < 0 {
                // Everything between the two ends is close enough to the line between them to go
                continue
            }
            kept[farthest] = true
            ranges.append((first: range.first, last: farthest))
            ranges.append((first: farthest, last: range.last))
        }
        return points.enumerated().filter { kept[$0.offset] }.map { $0.element }
    }
}

File: IsraelHiking.Web/android/app/src/main/java/com/mapeak/car/SpatialHelper.kt
package com.mapeak.car

import kotlin.math.abs
import kotlin.math.asin
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt
import org.maplibre.android.geometry.LatLng

/**
 * The spatial calculations the car experience measures routes with - the counterpart of
 * SpatialHelper on the web client, and of SpatialHelper.swift on iOS. Turf allocates a feature
 * per call, which a route with tens of thousands of points cannot afford when it is re-measured
 * against the GPS position on every fix.
 */
object SpatialHelper {

    /** Good enough for the local, flat plane calculations, the earth is not a perfect sphere anyway. */
    const val METERS_PER_LATITUDE_DEGREE = 111_320.0

    private const val EARTH_RADIUS_METERS = 6_371_008.8
    private const val DEGREES_TO_RADIANS = Math.PI / 180.0

    /** How many meters a degree of longitude is worth at the given latitude. */
    fun metersPerLongitudeDegree(latitude: Double): Double =
            METERS_PER_LATITUDE_DEGREE * cos(latitude * DEGREES_TO_RADIANS)

    /**
     * The distance in meters between two positions, measured on the sphere. Used for the distances
     * along a route, which are measured once per route and have to hold over its whole length.
     */
    fun distanceMeters(from: LatLng, to: LatLng): Double {
        val fromLatitude = from.latitude * DEGREES_TO_RADIANS
        val toLatitude = to.latitude * DEGREES_TO_RADIANS
        val latitudeDelta = (to.latitude - from.latitude) * DEGREES_TO_RADIANS
        val longitudeDelta = (to.longitude - from.longitude) * DEGREES_TO_RADIANS
        val a =
                sin(latitudeDelta / 2) * sin(latitudeDelta / 2) +
                        cos(fromLatitude) *
                                cos(toLatitude) *
                                sin(longitudeDelta / 2) *
                                sin(longitudeDelta / 2)
        return 2 * EARTH_RADIUS_METERS * asin(sqrt(a).coerceAtMost(1.0))
    }

    /** The bearing in degrees within [0, 360) from one position to another. */
    fun bearingDegrees(from: LatLng, to: LatLng): Double {
        val fromLatitude = from.latitude * DEGREES_TO_RADIANS
        val toLatitude = to.latitude * DEGREES_TO_RADIANS
        val longitudeDelta = (to.longitude - from.longitude) * DEGREES_TO_RADIANS
        val y = sin(longitudeDelta) * cos(toLatitude)
        val x =
                cos(fromLatitude) * sin(toLatitude) -
                        sin(fromLatitude) * cos(toLatitude) * cos(longitudeDelta)
        return (Math.toDegrees(Math.atan2(y, x)) + 360.0) % 360.0
    }

    /** Smallest absolute difference between two bearings, in degrees within [0, 180]. */
    fun angleDifference(a: Double, b: Double): Double {
        val difference = abs(a - b) % 360.0
        return if (difference > 180.0) 360.0 - difference else difference
    }

    /**
     * The distance in meters between a point and the closest place on a segment, measured on a
     * plane laid around the point.
     */
    fun perpendicularDistanceMeters(point: LatLng, from: LatLng, to: LatLng): Double {
        val metersPerLongitudeDegree = metersPerLongitudeDegree(point.latitude)
        val startX = (from.longitude - point.longitude) * metersPerLongitudeDegree
        val startY = (from.latitude - point.latitude) * METERS_PER_LATITUDE_DEGREE
        val endX = (to.longitude - point.longitude) * metersPerLongitudeDegree
        val endY = (to.latitude - point.latitude) * METERS_PER_LATITUDE_DEGREE
        val deltaX = endX - startX
        val deltaY = endY - startY
        val lengthSquared = deltaX * deltaX + deltaY * deltaY
        val projectionFactor =
                if (lengthSquared > 0)
                        (-(startX * deltaX + startY * deltaY) / lengthSquared).coerceIn(0.0, 1.0)
                else 0.0
        val x = startX + projectionFactor * deltaX
        val y = startY + projectionFactor * deltaY
        return sqrt(x * x + y * y)
    }

    /**
     * Drops the points that fall within [toleranceMeters] of the line their neighbours draw
     * (Ramer-Douglas-Peucker), keeping the corners. Iterative rather than recursive, since a
     * recorded route can hold tens of thousands of points.
     */
    fun simplify(points: List<LatLng>, toleranceMeters: Double): List<LatLng> {
        if (points.size <= 2) {
            return points
        }
        val kept = BooleanArray(points.size)
        kept[0] = true
        kept[points.size - 1] = true
        val ranges = ArrayDeque<Pair<Int, Int>>()
        ranges.addLast(0 to points.size - 1)
        while (ranges.isNotEmpty()) {
            val (first, last) = ranges.removeLast()
            if (last <= first + 1) {
                continue
            }
            var farthest = -1
            var farthestDistance = toleranceMeters
            for (index in first + 1 until last) {
                val distance =
                        perpendicularDistanceMeters(points[index], points[first], points[last])
                if (distance > farthestDistance) {
                    farthestDistance = distance
                    farthest = index
                }
            }
            if (farthest < 0) {
                // Everything between the two ends is close enough to the line between them to go
                continue
            }
            kept[farthest] = true
            ranges.addLast(first to farthest)
            ranges.addLast(farthest to last)
        }
        return points.filterIndexed { index, _ -> kept[index] }
    }
}
